internal/graph/lookup.go

```diff
@@ -43,864 +38,97 @@ type ValidatedLookupRequest struct {
-		}
-
-		toCheck = tuple.NewONRSet()
-
-		for _, obj := range result.Resp.ResolvedOnrs {
-			// Only check recursively for new objects.
-			if objSet.Add(obj) {
-				toCheck.Add(obj)
-			}
-		}
-	}
-
-	return returnResult(lookupResult(req, limitedSlice(objSet.AsSlice(), req.Limit), responseMetadata))
+	return nil
 }
 
-func (cl *ConcurrentLookup) lookupDirect(ctx context.Context, req ValidatedLookupRequest, typeSystem *namespace.TypeSystem) ReduceableLookupFunc {
-	requests := []ReduceableLookupFunc{}
-
-	// Ensure type informatione exists on the relation.
-	if !typeSystem.HasTypeInformation(req.ObjectRelation.Relation) {
-		return returnResult(lookupResultError(
-			req,
-			NewRelationMissingTypeInfoErr(req.ObjectRelation.Namespace, req.ObjectRelation.Relation),
-			emptyMetadata,
-		))
-	}
-
-	// Dispatch a check for the target ONR directly, if it is allowed on the start relation.
-	isDirectAllowed, err := typeSystem.IsAllowedDirectRelation(req.ObjectRelation.Relation, req.Subject.Namespace, req.Subject.Relation)
-	if err != nil {
-		return returnResult(lookupResultError(req, err, emptyMetadata))
+func (cl *ConcurrentLookup) LookupViaReachability(ctx context.Context, req ValidatedLookupRequest) (*v1.DispatchLookupResponse, error) {
+	if req.Subject.ObjectId == tuple.PublicWildcard {
+		resp := lookupResultError(req, NewErrInvalidArgument(errors.New("cannot perform lookup on wildcard")), emptyMetadata)
+		return resp.Resp, resp.Err
 	}
 
-	ds := datastoremw.MustFromContext(ctx)
-
-	if isDirectAllowed == namespace.DirectRelationValid {
-		requests = append(requests, func(ctx context.Context, resultChan chan<- LookupResult) {
-			objects := tuple.NewONRSet()
-			it, err := ds.ReverseQueryTuples(
-				ctx,
-				tuple.UsersetToSubjectFilter(req.Subject),
-				req.Revision,
-				options.WithResRelation(&options.ResourceRelation{
-					Namespace: req.ObjectRelation.Namespace,
-					Relation:  req.ObjectRelation.Relation,
-				}),
-			)
-			if err != nil {
-				resultChan <- lookupResultError(req, err, emptyMetadata)
-				return
-			}
-			defer it.Close()
-
-			for tpl := it.Next(); tpl != nil; tpl = it.Next() {
-				objects.Add(tpl.ObjectAndRelation)
-				if objects.Length() >= req.Limit {
-					break
-				}
-			}
+	cancelCtx, checkCancel := context.WithCancel(ctx)
+	defer checkCancel()
 
-			if it.Err() != nil {
-				resultChan <- lookupResultError(req, it.Err(), emptyMetadata)
-				return
-			}
+	checker := NewParallelChecker(cancelCtx, cl.c, req.Subject, 10)
+	stream := &collectingStream{checker, req, cancelCtx, 0, 0, 0, sync.Mutex{}}
 
-			resultChan <- lookupResult(req, objects.AsSlice(), emptyMetadata)
-		})
-	}
+	// Start the checker.
+	checker.Start()
 
-	// Dispatch a check for the subject wildcard, if allowed.
-	isWildcardAllowed, err := typeSystem.IsAllowedPublicNamespace(req.ObjectRelation.Relation, req.Subject.Namespace)
+	// TODO(jschorr): Collect dispatch count and depth required.
```

Comment:
What's the plan for reporting this? Add the count to the lookup count and take the max depth?